### PR TITLE
[CARBONDATA-816]  Added Example for Hive Integration

### PR DESCRIPTION
**Pull request by anubhav100**
Added Example For HIve Carbon Integration

### CODE REVIEW

**Comment by anubhav100** [comment]
@chenliang613 can you please have a look?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1490/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1495/

File: integration/hive/pom.xml [hunk 1]
```diff
@@ -95,7 +168,9 @@
                 <!-- Note config is repeated in scalatest config -->
                 <configuration>
                     <includes>
-                        <include>**/Test*.java</include>
```
Comment (QiangCai) [inline integration/hive/pom.xml]: add maven-scala-plugin, like spark2 module

File: integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    import org.apache.spark.sql.CarbonSession._
+
+    val carbon = SparkSession
+      .builder()
+      .master("local")
+      .appName("HiveExample")
+      .config("carbon.sql.warehouse.dir", warehouse).enableHiveSupport()
+      .getOrCreateCarbonSession(
+        "hdfs://localhost:54310/opt/carbonStore", metaStore_Db)
```
Comment (QiangCai) [inline integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala]: change to use local dir

File: integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    val hiveEmbeddedServer2 = new HiveEmbeddedServer2()
+    hiveEmbeddedServer2.start()
+    val port = hiveEmbeddedServer2.getFreePort
+    val con = DriverManager.getConnection(s"jdbc:hive2://localhost:$port/default", "", "")
+    val stmt: Statement = con.createStatement
+
+    logger.info(s"============HIVE CLI IS STARTED ON PORT $port ==============")
+
+    stmt
+      .execute(s"ADD JAR $rootPath/assembly/target/scala-2.11/carbondata_2.11-1.1" +
```
Comment (QiangCai) [inline integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala]: check whether jar file is exists or not

File: integration/hive/pom.xml (hunk 1 above)
Comment (anubhav100) [inline integration/hive/pom.xml]: changes done

File: integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala (hunk 2 above)
Comment (anubhav100) [inline integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala]: @QiangCai changes done

File: integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala (hunk 3 above)
Comment (anubhav100) [inline integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala]: @QiangCai  changes done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1518/

**Comment by anubhav100** [comment]
@chenliang613 @QiangCai changes done can you have a look?

File: integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    val carbon = SparkSession
+      .builder()
+      .master("local")
+      .appName("HiveExample")
+      .config("carbon.sql.warehouse.dir", warehouse).enableHiveSupport()
+      .getOrCreateCarbonSession(
+        store, metaStore_Db)
+
+    val carbonJarPath = s"$rootPath/assembly/target/scala-2.11/carbondata_2.11-1.1" +
+                        s".0-incubating-SNAPSHOT-shade-hadoop2.7.2.jar"
```
Comment (chenliang613) [inline integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala]: For hadoop version of assembly jar , please don't give the fixed version number(2.7.2), how about using 2.* ?

Comment (anubhav100) [inline integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala]: @chenliang613  Wildcards are not supported in both hive.aux.jars.path and add jar.

i solved this problem by specifying both version of hadoop

val carbonHadoopJarPath = s"$rootPath/assembly/target/scala-2.11/carbondata_2.11-1.1" +
                              ".0-incubating-SNAPSHOT-shade-hadoop2.7.2.jar"

    val carbon_DefaultHadoopVersion_JarPath =
      s"$rootPath/assembly/target/scala-2.11/carbondata_2.11-1.1" +
      ".0-incubating-SNAPSHOT-shade-hadoop2.2.0.jar"

first i tried to add jar with 2.7.2 version if it fails then i tried with hadoop 2.2.0 version

 try {
      stmt
        .execute(s"ADD JAR $carbonHadoopJarPath")
    }
    catch {
      case exception: Exception =>
        logger.warn(s"Jar Not Found $carbonHadoopJarPath"+"Looking For hadoop 2.2.0 version jar")
        try {
          stmt
            .execute(s"ADD JAR $carbon_DefaultHadoopVersion_JarPath")
        }
        catch {
          case exception: Exception => logger
            .error(s"Exception Occurs:Neither One of Jar is Found $carbon_DefaultHadoopVersion_JarPath,$carbonHadoopJarPath"+"Atleast One Should Be Build")
            hiveEmbeddedServer2.stop()
        }
    }

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1529/

**Comment by cenyuhai** [comment]
@anubhav100 can you create a new project under examples and move these codes?
examples
------flink
------spark
------spark2
------hive
---------your codes

File: integration/hive/pom.xml [hunk 5]
```diff
@@ -64,6 +64,79 @@
...
+            <artifactId>httpclient</artifactId>
+            <version>4.3.4</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpcore</artifactId>
+            <version>4.3-alpha1</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
```
Comment (cenyuhai) [inline integration/hive/pom.xml]: hadoop-client already contains hadoop-hdfs and hadoop-common, what hadoop version do you want?

**Comment by chenliang613** [comment]
@cenyuhai  HiveExample can more to examples/ later once hive integration module be completely moved into master.

File: integration/hive/pom.xml [hunk 6]
```diff
@@ -64,6 +64,79 @@
             <scope>compile</scope>
         </dependency>
         <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-network-common_2.10</artifactId>
+            <version>2.1.0</version>
```
Comment (cenyuhai) [inline integration/hive/pom.xml]: carbondata-spark2 has already included this dependency

File: integration/hive/pom.xml [hunk 7]
```diff
@@ -64,6 +64,79 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-hdfs</artifactId>
+            <version>2.7.3</version>
```
Comment (cenyuhai) [inline integration/hive/pom.xml]: it is already included in carbondata-hadoop

File: integration/hive/pom.xml [hunk 8]
```diff
@@ -64,6 +64,79 @@
...
+            <artifactId>spark-sql_${scala.binary.version}</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-hive-thriftserver_${scala.binary.version}</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-repl_${scala.binary.version}</artifactId>
+        </dependency>
```
Comment (cenyuhai) [inline integration/hive/pom.xml]: it is already included in carbondata-spark2

**Comment by cenyuhai** [comment]
I test this example in my computer, it's very good. But I think you should remove some unnecessary dependencies.

File: integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+      ".0-incubating-SNAPSHOT-shade-hadoop2.2.0.jar"
+
+    val hiveJarPath = s"$rootPath/integration/hive/target/carbondata-hive-1.1" +
+                      ".0-incubating-SNAPSHOT.jar"
+
+    carbon.sql("""drop table if exists hive_carbon_example""".stripMargin)
+
+    carbon
+      .sql(
+        """create table hive_carbon_example (id int,name string,salary double) stored by
```
Comment (chenliang613) [inline integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala]: Please use uppercase for DDL and DML key words.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1553/

File: integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala (hunk 9 above)
Comment (anubhav100) [inline integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala]: @chenliang613 resolved

File: integration/hive/pom.xml (hunk 8 above)
Comment (anubhav100) [inline integration/hive/pom.xml]: @cenyuhai resolved comments

**Comment by anubhav100** [comment]
@cenyuhai @chenliang613 i resolved all the comments can please have a look?can it be merged?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1557/

File: integration/hive/pom.xml (hunk 5 above)
Comment (anubhav100) [inline integration/hive/pom.xml]: done

File: integration/hive/pom.xml (hunk 6 above)
Comment (anubhav100) [inline integration/hive/pom.xml]: resolved

File: integration/hive/pom.xml (hunk 7 above)
Comment (anubhav100) [inline integration/hive/pom.xml]: resolved

**Comment by chenliang613** [comment]
@anubhav100  everything looks good, can you squash all commits into one commit.

When ending the HiveExample, the application can't be automatically stopped, we must to manually stop it:

**Comment by anubhav100** [comment]
@chenliang613 i added the code for stoping application automatically ,tested it on my system and rebased all commits into one commit can it be merged?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1579/

**Comment by chenliang613** [comment]
@anubhav100  Thanks for your good contribution.

LGTM

**Pull request closed without merging**